bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,80 +1,302 @@
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of retries the worker has had.
+    pub fn num_retries(&self) -> usize {
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => 0,
+            WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts,
+                first_fail_timestamp: _,
+                reason: _,
+            } => *failed_attempts,
+            WorkerHealthStatus::Unknown => 0,
+        }
+    }
+
+    /// Returns the health status of the worker.
+    pub fn health_status(&self) -> &WorkerHealthStatus {
+        &self.health_status
+    }
+
+    /// Returns the version of the worker.
+    pub fn version(&self) -> &str {
+        &self.version
+    }
+
+    /// Returns the worker availability.
+    pub fn is_available(&self) -> bool {
+        self.is_available
+    }
+
+    /// Returns the worker address.
     pub fn address(&self) -> String {
         self.backend.addr.to_string()
     }
 
-    pub async fn is_healthy(&mut self) -> bool {
-        match self
-            .health_check_client
-            .check(HealthCheckRequest { service: "".to_string() })
-            .await
-        {
-            Ok(response) => response.into_inner().status() == ServingStatus::Serving,
-            Err(err) => {
-                error!("Failed to check worker health ({}): {}", self.address(), err);
-                false
+    // PRIVATE HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns whether the worker should do a health check.
+    ///
+    /// A worker should do a health check if it is healthy or if the time since the first failure
+    /// is greater than the time since the first failure power of 2.
+    ///
+    /// The maximum exponent is [MAX_BACKOFF_EXPONENT], which corresponds to a backoff of
+    /// 2^[MAX_BACKOFF_EXPONENT] seconds.
+    fn should_do_health_check(&self) -> bool {
+        match self.health_status {
+            WorkerHealthStatus::Healthy => true,
+            WorkerHealthStatus::Unhealthy {
+                num_failed_attempts: failed_attempts,
+                first_fail_timestamp,
+                reason: _,
+            } => {
+                let time_since_first_failure = Instant::now() - first_fail_timestamp;
+                time_since_first_failure
+                    > Duration::from_secs(
+                        2u64.pow(failed_attempts.min(MAX_BACKOFF_EXPONENT) as u32),
+                    )
             },
+            WorkerHealthStatus::Unknown => true,
         }
     }
 
-    pub fn is_available(&self) -> bool {
-        self.is_available
+    /// Sets the health status of the worker.
+    fn set_health_status(&mut self, health_status: WorkerHealthStatus) {
+        let previous_health_status = self.health_status.clone();
```

Comment:
nit: I would maybe add `WorkerHealthStatus::is_healthy()` method and then would change this to:
```rust
let was_healthy = self.health_status.is_healthy();
```
This should simplify the code below a bit.